File: src/go/core/internal/render/options.go
package render

import (
	"strings"

	"github.com/dicebear/dicebear-go/v10/internal/prng"
)

// tagFilterToken is a parsed `tags` filter token. options.tags decodes each raw
// `category` / `category:value` / `!…` string into this shape so the resolver
// composes the filter without parsing the grammar itself. hasValue distinguishes
// a bare category (`cat`) from an empty value (`cat:`).
type tagFilterToken struct {
	category string
	value    string
	hasValue bool
	negated  bool
}

// options reads and normalizes the raw user-supplied options. Each accessor
// returns the user's input in a normalized form (always a slice for options
// that accept a scalar or a slice, or the zero/absent form when unset), so the
// resolver never has to normalize. The data map is the options object after a
// JSON round-trip, so every number is a float64 and every array a []any.
type options struct {
	data       map[string]any
	tagsParsed bool
	tagTokens  []tagFilterToken
}

func newOptions(data map[string]any) *options {
	if data == nil {
		data = map[string]any{}
	}
	return &options{data: data}
}

func (o *options) get(key string) (any, bool) {
	v, ok := o.data[key]
	if !ok || v == nil {
		return nil, false
	}
	return v, true
}

func (o *options) seed() (string, bool) {
	if v, ok := o.get("seed"); ok {
		if s, ok := v.(string); ok {
			return s, true
		}
	}
	return "", false
}

func (o *options) size() *float64 {
	if v, ok := o.get("size"); ok {
		if n, ok := v.(float64); ok {
			return &n
		}
	}
	return nil
}

func (o *options) idRandomization() *bool {
	if v, ok := o.get("idRandomization"); ok {
		if b, ok := v.(bool); ok {
			return &b
		}
	}
	return nil
}

func (o *options) title() (string, bool) {
	if v, ok := o.get("title"); ok {
		if s, ok := v.(string); ok {
			return s, true
		}
	}
	return "", false
}

func (o *options) flip() []string        { v, _ := o.get("flip"); return asStringArray(v) }
func (o *options) fontFamily() []string  { v, _ := o.get("fontFamily"); return asStringArray(v) }
func (o *options) fontWeight() []float64 { v, _ := o.get("fontWeight"); return asNumberArray(v) }

func (o *options) scale() *prng.Range        { v, _ := o.get("scale"); return toRange(v) }
func (o *options) borderRadius() *prng.Range { v, _ := o.get("borderRadius"); return toRange(v) }
func (o *options) rotate() *prng.Range       { v, _ := o.get("rotate"); return toRange(v) }
func (o *options) translateX() *prng.Range   { v, _ := o.get("translateX"); return toRange(v) }
func (o *options) translateY() *prng.Range   { v, _ := o.get("translateY"); return toRange(v) }

// tags returns the global `tags` filter as parsed tokens, or an empty slice when
// unset. Each raw token (`category` / `category:value`, optionally `!`-prefixed
// to exclude) is decoded into a tagFilterToken so the resolver composes the
// filter without parsing the grammar itself. Memoized, since the resolver reads
// it once per component.
func (o *options) tags() []tagFilterToken {
	if o.tagsParsed {
		return o.tagTokens
	}
	o.tagsParsed = true

	v, _ := o.get("tags")
	raw := asStringArray(v)
	tokens := make([]tagFilterToken, 0, len(raw))
	for _, token := range raw {
		negated := strings.HasPrefix(token, "!")
		body := token
		if negated {
			body = token[1:]
		}

		if sep := strings.IndexByte(body, ':'); sep == -1 {
			tokens = append(tokens, tagFilterToken{category: body, negated: negated})
		} else {
			tokens = append(tokens, tagFilterToken{
				category: body[:sep],
				value:    body[sep+1:],
				hasValue: true,
				negated:  negated,
			})
		}
	}

	o.tagTokens = tokens
	return tokens
}

// componentVariant returns the ${name}Variant constraint as a weighted map, or
// (nil, false) when unset. A bare string or string list is normalized to weight
// 1 each.
func (o *options) componentVariant(name string) (map[string]float64, bool) {
	raw, ok := o.get(name + "Variant")
	if !ok {
		return nil, false
	}

	switch val := raw.(type) {
	case string:
		return map[string]float64{val: 1}, true
	case []any:
		out := map[string]float64{}
		for _, item := range val {
			if s, ok := item.(string); ok {
				out[s] = 1
			}
		}
		return out, true
	case map[string]any:
		out := map[string]float64{}
		for k, v := range val {
			if w, ok := v.(float64); ok {
				out[k] = w
			}
		}
		return out, true
	default:
		return nil, false
	}
}

func (o *options) componentProbability(name string) *float64 {
	if v, ok := o.get(name + "Probability"); ok {
		if n, ok := v.(float64); ok {
			return &n
		}
	}
	return nil
}

// color returns (nil, false) when ${name}Color is unset, so the resolver can
// fall back to the style definition's values.
func (o *options) color(name string) ([]string, bool) {
	raw, ok := o.get(name + "Color")
	if !ok {
		return nil, false
	}
	return asStringArray(raw), true
}

func (o *options) colorFill(name string) []string {
	v, _ := o.get(name + "ColorFill")
	return asStringArray(v)
}

func (o *options) colorAngle(name string) *prng.Range {
	v, _ := o.get(name + "ColorAngle")
	return toRange(v)
}

func (o *options) colorFillStops(name string) *prng.Range {
	v, _ := o.get(name + "ColorFillStops")
	return toRange(v)
}

// asStringArray normalizes a scalar/array/absent string value into a slice.
func asStringArray(value any) []string {
	switch v := value.(type) {
	case []any:
		out := make([]string, 0, len(v))
		for _, item := range v {
			if s, ok := item.(string); ok {
				out = append(out, s)
			}
		}
		return out
	case string:
		return []string{v}
	default:
		return nil
	}
}

// asNumberArray normalizes a scalar/array/absent numeric value into a slice.
func asNumberArray(value any) []float64 {
	switch v := value.(type) {
	case []any:
		out := make([]float64, 0, len(v))
		for _, item := range v {
			if n, ok := item.(float64); ok {
				out = append(out, n)
			}
		}
		return out
	case float64:
		return []float64{v}
	default:
		return nil
	}
}

// toRange normalizes a range option (bare number, [n], [min, max], or absent)
// into a *prng.Range. A bare number becomes a fixed min == max; an empty array
// is treated as unset. Matches the JS, PHP, Python, Rust and Dart ports.
func toRange(value any) *prng.Range {
	switch v := value.(type) {
	case float64:
		return &prng.Range{Min: v, Max: v}
	case []any:
		var nums []float64
		for _, item := range v {
			if n, ok := item.(float64); ok {
				nums = append(nums, n)
			}
		}
		if len(nums) == 0 {
			return nil
		}
		min, max := nums[0], nums[0]
		for _, n := range nums {
			if n < min {
				min = n
			}
			if n > max {
				max = n
			}
		}
		return &prng.Range{Min: min, Max: max}
	default:
		return nil
	}
}

File: src/rust/core/src/options.rs
//! Reads and normalizes the raw user-supplied options.
//!
//! Each accessor returns the user's input in a normalized form (always a list
//! for options that accept a scalar or a list, or `None` when unset), so the
//! [`crate::Resolver`] never has to normalize. Resolution against the style and
//! the PRNG happens there; this type is purely about reading input.

use std::cell::OnceCell;
use std::collections::HashMap;

use serde_json::Value;

use crate::prng::Range;

/// The two values of the per-color `${name}ColorOrder` option.
pub(crate) const COLOR_ORDER_RANDOM: &str = "random";
pub(crate) const COLOR_ORDER_FIXED: &str = "fixed";

/// A parsed `tags` filter token. [`Options::tags`] decodes each raw
/// `category` / `category:value` / `!…` string into this shape so the resolver
/// composes the filter without parsing the grammar itself.
pub(crate) struct TagFilterToken {
    pub category: String,
    pub value: Option<String>,
    pub negated: bool,
}

/// The interpreted `animation` option: a boolean switch, or a selection of
/// timeline names in user order (a bare name is normalized to a one-element
/// list). Keeps the raw shape so the resolved-options snapshot serializes the
/// user's input, while consumers work with the accessors.
#[derive(Clone)]
pub(crate) enum AnimationSelection {
    Flag(bool),
    Names(Vec<String>),
}

impl AnimationSelection {
    /// Whether no timeline plays at all.
    pub(crate) fn off(&self) -> bool {
        match self {
            Self::Flag(flag) => !flag,
            Self::Names(names) => names.is_empty(),
        }
    }

    /// Whether a timeline carrying the given name plays. `true` plays every
    /// timeline; a name selection plays only named timelines carrying one of
    /// the selected names.
    pub(crate) fn matches(&self, name: Option<&str>) -> bool {
        match self {
            Self::Flag(flag) => *flag,
            Self::Names(names) => {
                name.is_some_and(|name| names.iter().any(|candidate| candidate == name))
            }
        }
    }

    /// The selected names in user order, or `None` for the boolean forms.
    /// Consumed by the animation class namespace hash.
    pub(crate) fn names(&self) -> Option<&[String]> {
        match self {
            Self::Flag(_) => None,
            Self::Names(names) => Some(names),
        }
    }

    /// The raw option value for the resolved-options snapshot: the boolean as
    /// given, or the name list in user order as a JSON array.
    pub(crate) fn to_value(&self) -> Value {
        match self {
            Self::Flag(flag) => Value::from(*flag),
            Self::Names(names) => Value::from(names.clone()),
        }
    }
}

pub struct Options {
    data: Value,
    tags: OnceCell<Vec<TagFilterToken>>,
}

impl Options {
    pub fn new(data: Value) -> Self {
        let data = if data.is_object() {
            data
        } else {
            Value::Object(serde_json::Map::new())
        };

        Self {
            data,
            tags: OnceCell::new(),
        }
    }

    fn get(&self, key: &str) -> Option<&Value> {
        self.data.get(key).filter(|v| !v.is_null())
    }

    pub fn seed(&self) -> Option<String> {
        self.get("seed").and_then(Value::as_str).map(String::from)
    }

    pub fn size(&self) -> Option<f64> {
        self.get("size").and_then(Value::as_f64)
    }

    pub fn id_randomization(&self) -> Option<bool> {
        self.get("idRandomization").and_then(Value::as_bool)
    }

    pub fn title(&self) -> Option<String> {
        self.get("title").and_then(Value::as_str).map(String::from)
    }

    pub fn flip(&self) -> Vec<String> {
        as_string_array(self.get("flip"))
    }

    pub fn font_family(&self) -> Vec<String> {
        as_string_array(self.get("fontFamily"))
    }

    pub fn font_weight(&self) -> Vec<f64> {
        as_number_array(self.get("fontWeight"))
    }

    pub fn scale(&self) -> Option<Range> {
        to_range(self.get("scale"))
    }

    pub fn border_radius(&self) -> Option<Range> {
        to_range(self.get("borderRadius"))
    }

    pub fn rotate(&self) -> Option<Range> {
        to_range(self.get("rotate"))
    }

    pub fn translate_x(&self) -> Option<Range> {
        to_range(self.get("translateX"))
    }

    pub fn translate_y(&self) -> Option<Range> {
        to_range(self.get("translateY"))
    }

    /// Returns the animation switch as given for booleans, or a name
    /// selection normalized to a list (a bare name becomes a one-element
    /// list). `None` when unset.
    pub(crate) fn animation(&self) -> Option<AnimationSelection> {
        match self.get("animation")? {
            Value::Bool(flag) => Some(AnimationSelection::Flag(*flag)),
            value => Some(AnimationSelection::Names(as_string_array(Some(value)))),
        }
    }

    pub fn animation_speed(&self) -> Option<Range> {
        to_range(self.get("animationSpeed"))
    }

    /// Returns the global `tags` filter as parsed tokens, or an empty list when
    /// unset. Each raw token (`category` / `category:value`, optionally
    /// `!`-prefixed to disallow) is decoded into `{ category, value?, negated }`
    /// so the resolver composes the filter without parsing the grammar itself.
    /// An empty list means no tag filtering (classic behavior). Memoized, since
    /// the resolver reads it once per component.
    pub(crate) fn tags(&self) -> &[TagFilterToken] {
        self.tags.get_or_init(|| {
            as_string_array(self.get("tags"))
                .into_iter()
                .map(|token| {
                    let negated = token.starts_with('!');
                    let body = if negated { &token[1..] } else { &token[..] };

                    match body.find(':') {
                        None => TagFilterToken {
                            category: body.to_string(),
                            value: None,
                            negated,
                        },
                        Some(sep) => TagFilterToken {
                            category: body[..sep].to_string(),
                            value: Some(body[sep + 1..].to_string()),
                            negated,
                        },
                    }
                })
                .collect()
        })
    }

    /// Returns the `${name}Variant` constraint as a weighted map, or `None` when
    /// unset. A bare string or string list is normalized to weight `1` each.
    pub fn component_variant(&self, name: &str) -> Option<HashMap<String, f64>> {
        match self.get(&format!("{name}Variant"))? {
            Value::String(s) => Some(HashMap::from([(s.clone(), 1.0)])),
            Value::Array(a) => Some(
                a.iter()
                    .filter_map(Value::as_str)
                    .map(|s| (s.to_string(), 1.0))
                    .collect(),
            ),
            Value::Object(o) => Some(
                o.iter()
                    .filter_map(|(k, v)| v.as_f64().map(|w| (k.clone(), w)))
                    .collect(),
            ),
            _ => None,
        }
    }

    pub fn component_probability(&self, name: &str) -> Option<f64> {
        self.get(&format!("{name}Probability"))
            .and_then(Value::as_f64)
    }

    /// Returns `None` (rather than `[]`) when `${name}Color` is unset, so the
    /// resolver can fall back to the style definition's values.
    pub fn color(&self, name: &str) -> Option<Vec<String>> {
        self.get(&format!("{name}Color"))
            .map(|v| as_string_array(Some(v)))
    }

    pub fn color_fill(&self, name: &str) -> Vec<String> {
        as_string_array(self.get(&format!("{name}ColorFill")))
    }

    pub fn color_angle(&self, name: &str) -> Option<Range> {
        to_range(self.get(&format!("{name}ColorAngle")))
    }

    pub fn color_fill_stops(&self, name: &str) -> Option<Range> {
        to_range(self.get(&format!("{name}ColorFillStops")))
    }

    pub fn color_order(&self, name: &str) -> Option<String> {
        self.get(&format!("{name}ColorOrder"))
            .and_then(Value::as_str)
            .map(String::from)
    }
}

/// Normalizes a scalar/array/absent string value into a list.
fn as_string_array(value: Option<&Value>) -> Vec<String> {
    match value {
        Some(Value::Array(a)) => a
            .iter()
            .filter_map(Value::as_str)
            .map(String::from)
            .collect(),
        Some(Value::String(s)) => vec![s.clone()],
        _ => Vec::new(),
    }
}

/// Normalizes a scalar/array/absent numeric value into a list.
fn as_number_array(value: Option<&Value>) -> Vec<f64> {
    match value {
        Some(Value::Array(a)) => a.iter().filter_map(Value::as_f64).collect(),
        Some(v) => v.as_f64().map(|n| vec![n]).unwrap_or_default(),
        None => Vec::new(),
    }
}

/// Normalizes a range option (bare number, `[n]`, `[min, max]`, or absent) into
/// a [`Range`]. A bare number becomes a fixed `min == max`; an empty array is
/// treated as unset. Matches every other port.
fn to_range(value: Option<&Value>) -> Option<Range> {
    match value {
        Some(Value::Number(n)) => {
            let x = n.as_f64()?;
            Some(Range {
                min: x,
                max: x,
                step: None,
            })
        }
        Some(Value::Array(a)) => {
            let nums: Vec<f64> = a.iter().filter_map(Value::as_f64).collect();
            if nums.is_empty() {
                return None;
            }
            let min = nums.iter().copied().fold(f64::INFINITY, f64::min);
            let max = nums.iter().copied().fold(f64::NEG_INFINITY, f64::max);
            Some(Range {
                min,
                max,
                step: None,
            })
        }
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::Options;

    #[test]
    fn color_order_returns_none_when_unset() {
        assert_eq!(Options::new(json!({})).color_order("skin"), None);
    }

    #[test]
    fn color_order_passes_the_value_through() {
        let options = Options::new(json!({ "skinColorOrder": "fixed" }));

        assert_eq!(options.color_order("skin").as_deref(), Some("fixed"));
    }
}
